### PR TITLE
Initial version of COMPARISON fulfillment

### PR DESCRIPTION
**Pull request by pradh**
This also re-uses the populate_charts() framework, but extends it to support multiple places.

Imminent TODO (hopefully tonight):  unit tests for nl_data_spec_next.py

Example: [auto theft in sunnyvale] -> [compare with palo alto]

Screenshot 2023-01-30 at 4 18 56 PM

### CODE REVIEW

File: server/lib/nl/nl_data_spec_next.py [hunk 1]
```diff
@@ -449,6 +498,30 @@ def _places_from_context(uttr: Utterance) -> List[Place]:
...
+      second_uttr_count = first_uttr_count + 1
+      while (second and second_uttr_count < CTX_LOOKBACK_LIMIT + 1):
+        # For the first place, add all combinations of the second.
+        if second.places:
+          ans.append(second.places + first.places)
+        second = second.prev_utterance
+        second_uttr_count = second_uttr_count + 1
+    elif len(first.places) > 1:
+      # Found multiple places in a single utterance.  Use this.
+      ans.append(first.places)
```
Comment (shifucun) [inline server/lib/nl/nl_data_spec_next.py:519]: do you want to break here?

**Review by shifucun — APPROVED** [review]
extending to multiple places is cool!

File: server/lib/nl/nl_data_spec_next.py (hunk 1 above)
Comment (pradh) [reply server/lib/nl/nl_data_spec_next.py:519]: We are returning multiple lists so the caller will try each related-places-list in order until successful.  So we don't want to break here?

Comment (shifucun) [reply server/lib/nl/nl_data_spec_next.py:519]: Can you add some comments or example for the choice here. For example for a query sequence of "tell me about san jose", "how is auto theft in fremont", "compare with palo alto". Will. the last one use all the places above (fremont, san jose)?

Comment (pradh) [reply server/lib/nl/nl_data_spec_next.py:539]: Clarified, please take a look.

File: server/lib/nl/nl_data_spec_next.py [hunk 2]
```diff
@@ -498,6 +498,26 @@ def _places_from_context(uttr: Utterance) -> List[Place]:
...
+# 1. If an utterance has multiple places, then that is a candidate list.
+# 2. If an utterance has a single place, then we walk up prior utterances until
+#    we find an utterance with at least one place.  The candidate list is
+#    the former single place plus the latter from prior utterance.
+#
+# For example, suppose the query sequence is as follows:
+#   [tell me about san jose] -> [how is auto theft in fremont] -> [compare with palo alto]
+#
+# In this case, the ordered candidates will be:
+#   [[palo alto, fremont], [palo alto, san jose], [fremont, san jose]]
```
Comment (shifucun) [inline server/lib/nl/nl_data_spec_next.py:514]: So we will show comparison for all the 3 combinations? Not sure if that will be too much, esp if there are more context places before.

**Comment by shifucun** [comment]
Thanks for adding examples!

File: server/lib/nl/nl_data_spec_next.py (hunk 2 above)
Comment (pradh) [reply server/lib/nl/nl_data_spec_next.py:514]: No, the caller will loop through each combo, and stop when the first one generated any charts.  (This is part of the general pattern of trying to show something when data doesn't exist for certain places)

Comment (shifucun) [reply server/lib/nl/nl_data_spec_next.py:514]: Ok, so here will fill as much as possible, now I see the pattern..

**Pull request merged**